### PR TITLE
Bind less when finding implicit object construction for FindReferences

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs [hunk 1]
```diff
@@ -623,34 +623,7 @@ static bool IsRelevantDocument(SyntaxTreeIndex syntaxTreeInfo)
             }
         }
 
-        protected Task<ImmutableArray<FinderLocation>> FindReferencesInImplicitObjectCreationExpressionAsync(
-            ISymbol symbol,
-            Document document,
-            SemanticModel semanticModel,
-            CancellationToken cancellationToken)
-        {
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs:631]: moved helper to only type it is used from

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs [hunk 2]
```diff
@@ -146,5 +147,55 @@ protected override bool CanFind(IMethodSymbol symbol)
...
+        {
+            // Only check `new (...)` calls that supply enough arguments to match all the required parameters for the constructor.
+            var minimumArgumentCount = symbol.Parameters.Count(p => !p.IsOptional && !p.IsParams);
+            var maximumArgumentCount = symbol.Parameters.Length > 0 && symbol.Parameters.Last().IsParams
+                ? int.MaxValue
+                : symbol.Parameters.Length;
+
+            var exactArgumentCount = symbol.Parameters.Any(p => p.IsOptional || p.IsParams)
+                ? -1
+                : symbol.Parameters.Length;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:165]: this is new code to help minimize the number of nodes to analyze.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs [hunk 3]
```diff
@@ -146,5 +147,55 @@ protected override bool CanFind(IMethodSymbol symbol)
...
+            static bool IsRelevantDocument(SyntaxTreeIndex syntaxTreeInfo)
+                => syntaxTreeInfo.ContainsImplicitObjectCreation;
+
+            void CollectMatchingReferences(
+                ISymbol originalUnreducedSymbolDefinition, SyntaxNode node,
+                ISyntaxFactsService syntaxFacts, ISemanticFactsService semanticFacts,
+                ArrayBuilder<FinderLocation> locations)
+            {
+                if (!syntaxFacts.IsImplicitObjectCreationExpression(node))
+                    return;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:178]: this is new code (@sharwell also added in https://github.com/dotnet/roslyn/pull/48401/files).

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs (hunk 1 above)
Comment (genlu) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs:631]: Moving and changing code in the same commit makes it hard to see the actual diff, I had to copy them into a diff tool to review.

**Review by genlu — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
> Moving and changing code in the same commit makes it hard to see the actual diff, I had to copy them into a diff tool to review.

Apologies :(

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs [hunk 4]
```diff
@@ -146,5 +147,55 @@ protected override bool CanFind(IMethodSymbol symbol)
...
+
+            static bool IsRelevantDocument(SyntaxTreeIndex syntaxTreeInfo)
+                => syntaxTreeInfo.ContainsImplicitObjectCreation;
+
+            void CollectMatchingReferences(
+                ISymbol originalUnreducedSymbolDefinition, SyntaxNode node,
+                ISyntaxFactsService syntaxFacts, ISemanticFactsService semanticFacts,
+                ArrayBuilder<FinderLocation> locations)
+            {
+                if (!syntaxFacts.IsImplicitObjectCreationExpression(node))
```
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:177]: @CyrusNajmabadi There is another instance method, "IsImplicitObjectCreation". Why we need to have an extension that does exactly the same thing?

**Review by Youssef1313 — COMMENTED** [review]
I can't see a test where the same document contains **both** target-typed new and the old object creation syntax.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs [hunk 5]
```diff
@@ -146,5 +147,55 @@ protected override bool CanFind(IMethodSymbol symbol)
...
+
+        private Task<ImmutableArray<FinderLocation>> FindReferencesInImplicitObjectCreationExpressionAsync(
+            IMethodSymbol symbol,
+            Document document,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            // Only check `new (...)` calls that supply enough arguments to match all the required parameters for the constructor.
+            var minimumArgumentCount = symbol.Parameters.Count(p => !p.IsOptional && !p.IsParams);
+            var maximumArgumentCount = symbol.Parameters.Length > 0 && symbol.Parameters.Last().IsParams
```
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:159]: Consider avoiding LINQ:

```suggestion
            var maximumArgumentCount = symbol.Parameters.Length > 0 && symbol.Parameters[^1].IsParams
```

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:159]: this should not be linq: https://docs.microsoft.com/en-us/dotnet/api/system.linq.immutablearrayextensions.last?view=netcore-3.1

**Pull request merged**